src/include/ipc/probe.h

```diff
@@ -94,7 +94,7 @@ struct probe_data_packet {
 	uint32_t checksum;		/**< CRC32 of header and payload */
 	uint32_t data_size_bytes;	/**< Size of following audio data */
 	uint32_t data[];		/**< Audio data extracted from buffer */
-} __attribute__((packed));
+} __attribute__((packed, aligned(4)));
```

Comment:
I don't like it being here. It shouldn't be concern of struct definition how aligned it is. Assumption is also bad, it just should be embedded in some local struct in .c file with `aligned` attribute. However I see others are already ok with this solution, so I won't be blocking over this.

Reply:
> Assumption is also bad

Which assumption?

> I don't like it being here. It shouldn't be concern of struct definition how aligned it is

I'm confused, where else could/should it be?

`((packed))` aligns every member to 1. The alignment of a struct is based on the alignment of its members:
https://gcc.gnu.org/onlinedocs/gcc-9.3.0/gcc/Common-Type-Attributes.html#Common-Type-Attributes
So `((packed))` is _already_ defining how this particular struct is (mis)aligned, which caused the warning that started all this. This `aligned(4)` is only undoing the struct alignment damage done by `((packed))` in the exact same place where the damage was done.

Even with this `aligned(4)` you can still override and (mis)align this struct anyway later if you want; for instance the following assert fails even with the `aligned(4)` above:
```diff
--- a/src/probe/probe.c
+++ b/src/probe/probe.c
@@ -69,9 +69,13 @@ struct probe_pdata {
 	struct probe_dma_ext ext_dma;				  /**< extraction DMA */
 	struct probe_dma_ext inject_dma[CONFIG_PROBE_DMA_MAX];	  /**< injection DMA */
 	struct probe_point probe_points[CONFIG_PROBE_POINTS_MAX]; /**< probe points */
+	uint8_t dummy;
+      // this local __packed overrides the more general struct aligned(4)
-	struct probe_data_packet header;           /**< data packet header */
+	struct probe_data_packet header __packed;  /**< data packet header */
 	struct task dmap_work;					  /**< probe task */
 };
+_Static_assert(offsetof(struct probe_pdata, header) % 4 == 0 ,
+               "probe_data_packet is not 4 aligned");
 ```

Speaking of undoing ((packed)) damage and if you really don't like `aligned(4)`,  I tested the following alternative which I think is 100% equivalent. It's verbose though:

```C
#define __packd __attribute__((packed))
struct probe_data_packet {
	uint32_t sync_word;   // first NOT PACKED	/**< PROBE_EXTRACT_SYNC_WORD */
	uint32_t buffer_id       __packd;		/**< Buffer ID from which data was extracted */
	uint32_t format         __packd;		/**< Encoded data format */
	uint32_t timestamp_low __packd;		/**< Low 32 bits of timestamp in us */
	uint32_t timestamp_high __packd;	/**< High 32 bits of timestamp in us */
	uint32_t checksum      __packd;		/**< CRC32 of header and payload */
	uint32_t data_size_bytes __packd;	/**< Size of following audio data */
	uint32_t data[]         __packd; /**< Audio data extracted from buffer */
};
```
All members packed except the first one.

https://gcc.gnu.org/onlinedocs/gcc-9.3.0/gcc/Common-Type-Attributes.html#Common-Type-Attributes

> Note that the alignment of any given struct or union type is required by the ISO C standard to be at least a perfect multiple of the lowest common multiple of the alignments of all of the members of the struct or union in question. This means that you can effectively adjust the alignment of a struct or union type by attaching an aligned attribute to any one of the members of such a type

Reply:
> > Assumption is also bad
> 
> Which assumption?

`__builtin_assume_aligned` this one ;p




> Even with this `aligned(4)` you can still override and (mis)align this struct anyway later if you want; for instance the following assert fails even with the `aligned(4)` above:
> 
> ```diff
> --- a/src/probe/probe.c
> +++ b/src/probe/probe.c
> @@ -69,9 +69,13 @@ struct probe_pdata {
>  	struct probe_dma_ext ext_dma;				  /**< extraction DMA */
>  	struct probe_dma_ext inject_dma[CONFIG_PROBE_DMA_MAX];	  /**< injection DMA */
>  	struct probe_point probe_points[CONFIG_PROBE_POINTS_MAX]; /**< probe points */
> +	uint8_t dummy;
> +      // this local __packed overrides the more general struct aligned(4)
> -	struct probe_data_packet header;           /**< data packet header */
> +	struct probe_data_packet header __packed;  /**< data packet header */
>  	struct task dmap_work;					  /**< probe task */
>  };
> +_Static_assert(offsetof(struct probe_pdata, header) % 4 == 0 ,
> +               "probe_data_packet is not 4 aligned");
> ```

Probably this is the place where I'd like to see it aligned instead of in struct def.
Something like:
```
struct probe_pdata {
 	struct probe_dma_ext ext_dma;				  /**< extraction DMA */
 	struct probe_dma_ext inject_dma[CONFIG_PROBE_DMA_MAX];	  /**< injection DMA */
 	struct probe_point probe_points[CONFIG_PROBE_POINTS_MAX]; /**< probe points */
 	struct probe_data_packet header __attribute__((aligned(4)));  /**< data packet header */
 	struct task dmap_work;					  /**< probe task */
 };
```

> `((packed))` aligns every member to 1. The alignment of a struct is based on the alignment of its members:
> https://gcc.gnu.org/onlinedocs/gcc-9.3.0/gcc/Common-Type-Attributes.html#Common-Type-Attributes
> So `((packed))` is _already_ defining how this particular struct is (mis)aligned, which caused the warning that started all this. This `aligned(4)` is only undoing the struct alignment damage done by `((packed))` in the exact same place where the damage was done.

I'm concerned about the fact that `packed` is only affecting layout within a struct, while `aligned` is affecting something outside it, imho if you really want to deal with alignment it should be defined in the place that is using this struct. Struct should define how it looks, not stuff around it. But these are just nitpicks ;)

Reply:
packed affects the layout but ALSO affects the minimum alignment of each members. Since all members are aligned at 1, the structure itself will be aligned at 1.

Structure alignment can be defined at the data type level because accessing a member of any struct should generate the same code, and you generate something to load an aligned value and something else (on at least some architectures, e.g. ARM) for misaligned values, because the hardware itself doesn't support misaligned loads (Intel does but most RISC architectures don't)

Reply:
> packed affects the layout but ALSO affects the minimum alignment of each members. Since all members are aligned at 1, the structure itself will be aligned at 1.

Well, yeah, you are right, about this one.
As I wrote I'm ok with merging this PR, but I'm still not convinced with this one issue ->


> Structure alignment can be defined at the data type level because accessing a member of any struct should generate the same code, and you generate something to load an aligned value and something else (on at least some architectures, e.g. ARM) for misaligned values, because the hardware itself doesn't support misaligned loads (Intel does but most RISC architectures don't)

Alignment at data type level in ABI is weird. It shouldn't matter for transfering bunch of bytes, especially when you have ipc between different archs. It's concern of receiver if it needs some alignment and should be defined on more *local* level.

Reply:
Guess this is a drawback of how C works. The alignment required at data type is, as I said, because different code is emitted to access aligned and misaligned words, on certain architectures. If you tried on ARM to access a misaligned word in assembler... Most likely you'd get either junk or the low bits of the address will be ignored (I didn't try though), unlike on Intel where it either works or you get an exception (depending on how the kernel configured it). Not sure what Xtensa says about unaligned loads or stores.

On Intel you can afford just emitting regular loads no matter the alignment. On ARM you can't, because that would be a correctness issue. And assuming all data is unaligned will f*ck performance with hidden issues where it isn't necessary.

We only use struct in the ABI to avoid having to manually fiddle with the bytes. And even this is vulnerable to endianness issues (thankfully Intel, ARM in the kernel and Xtensa are all little-endian, but how about big endian kernels [that'd be weird] or architectures that only support big endian?). This in itself is a hack, really. It's just a more commonly understood one.

Reply:
> This in itself is a hack, really.

Thanks @paulstelian97 , `__packed` is indeed a non standard hack. This took me a long time to realize considering how common and widely supported `__packed` is.

"C Is Not a Low-level Language - Your computer is not a fast PDP-11." https://queue.acm.org/detail.cfm?id=3212479 

> And even this is vulnerable to endianness issues 

I think the only sure way to serialize is to write each member one by one with `htonl()` into byte arrays and then send that.

 http://www.catb.org/esr/structure-packing/ is also interesting.

Reply:
Indeed. For now we will go with this solution. Maybe IPC 2.0 will do just that, serializing into byte arrays and using htonl/ntohl or similar. Probably something to propose.